ai_genomics/analysis/tag_evolution/tag_evolution_embeds.py

```diff
@@ -0,0 +1,244 @@
+    """
+    periods = 2022 - ast.literal_eval(start_date.split("-")[0])
+    date_range = pd.date_range(start=start_date, periods=periods, freq="A")
+
+    ents_per_date = dict()
+    for min_date in date_range:
+        all_ents = []
+        for df, ents in zip(list_of_dfs, list_of_ents):
+            df_ids = list(df[df.date < min_date].id)
+            ents_subset = {
+                id_: [i[0] for i in ents.get(id_)] for id_ in df_ids if ents.get(id_)
+            }
+            all_ents.extend(list(itertools.chain(*ents_subset.values())))
+        ents_per_date[min_date.year] = list(set(all_ents))
+
+    return ents_per_date
+
+
+def generate_entities_embedding_lookup(entities: List[str]) -> Dict[str, np.array]:
+    """Generates an embedding lookup where the key is the entity
+    and the value is the reduced embedding.
+    """
+    embeds = bert_model.transform(entities)
+    embeds_reduced = reducer.fit_transform(embeds)
+    ent_reduced_embeds_dict = dict(zip(entities, embeds_reduced))
+
+    return ent_reduced_embeds_dict
+
+
+# if this is the method we want to end up using, manually define k based on graphs
+# rather than max
+def get_best_k(
+    reduced_entities: Dict[str, np.array], ents_per_date: Dict[str, List[str]]
+) -> List[int]:
+    """Identifies the optimal number of clusters per timeslice based
+    on the maximum mean silouette score for k between 10-70 at 5 intervals.
+
+    Returns list of optimal ks per timeslice.
+    """
+    best_ks = []
+    for ents in ents_per_date.values():
+        results = []
+        ent_embeds = [reduced_entities.get(ent) for ent in ents]
+        for n in range(10, 70, 5):
+            km = KMeans(n_clusters=n)
+            clust = km.fit_predict(ent_embeds)
+            score = silhouette_score(ent_embeds, clust)
+            results.append([n, score])
+        max_score = max([i[1] for i in results])
+        best_ks.extend([i[0] for i in results if i[1] == max_score])
+
+    return best_ks
+
+
+def propagate_labels(
+    timeslice_x, timeslice_y, min_jaccard_score: float = 0.7,
+) -> Dict[int, List[str]]:
+    """Propogates cluster labels from t-1 to t based on minimum jaccard similarity
+    between entity lists.    
+    """
+    cluster_perms = list(itertools.product(timeslice_x, timeslice_y))
+
+    perm_dists = []
+    for cluster_x, cluster_y in cluster_perms:
+        timeslice_x_ents, timeslice_y_ents = (
+            timeslice_x.get(cluster_x),
+            timeslice_y.get(cluster_y),
+        )
+        dists = jaccard_similarity(timeslice_x_ents, timeslice_y_ents)
+        if (dists != 0) & (dists > min_jaccard_score):
+            perm_dists.append((cluster_x, cluster_y, dists))
+
+    sorted_perm_dists = sorted(perm_dists, key=lambda x: (x[0], x[2]), reverse=True)
+    sorted_perm_dists_clusts = list(
+        dict([(i[0], i[1]) for i in sorted_perm_dists]).items()
+    )
+
+    while len(sorted_perm_dists_clusts) > 0:
```

Comment:
I struggled to follow this function because it doesn't return anything. Are you modifying `timeslice_x` and `timeslice_y` in place? I think it would be nicer to return new modified objects. 

It might be easier to understand the logic if you change the name of arguments to e.g. `timeslice_t` and `timeslice_t_plus_1` or something like that.

Reply:
good idea- i've changed the variable names to be easier to understand and i've commented each line of code with a description of what's happening so hopefully its easier to follow